### PR TITLE
Initial Azure.ResourceManager.Workloads

### LINKED ISSUE #28848: Initial Azure.Resourcemanager.Workloads
https://github.com/Azure/azure-rest-api-specs/tree/main/specification/workloads/resource-manager

### PR DESCRIPTION
**Pull request by ArthurMa1978**

### CODE REVIEW

File: sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs [hunk 1]
```diff
@@ -0,0 +1,1742 @@
...
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public override bool Equals(object obj) { throw null; }
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public override int GetHashCode() { throw null; }
+        public static bool operator ==(Azure.ResourceManager.Workloads.Models.CentralServerVirtualMachineType left, Azure.ResourceManager.Workloads.Models.CentralServerVirtualMachineType right) { throw null; }
+        public static implicit operator Azure.ResourceManager.Workloads.Models.CentralServerVirtualMachineType (string value) { throw null; }
+        public static bool operator !=(Azure.ResourceManager.Workloads.Models.CentralServerVirtualMachineType left, Azure.ResourceManager.Workloads.Models.CentralServerVirtualMachineType right) { throw null; }
+        public override string ToString() { throw null; }
+    }
+    public partial class CentralServerVmDetails
```
Comment (fengzhou-msft) [inline sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs:515]: Should it be `CentralServerVMDetails`?

File: sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs [hunk 2]
```diff
@@ -0,0 +1,1742 @@
...
+    }
+    public partial class EnqueueServerProperties
+    {
+        public EnqueueServerProperties() { }
+        public Azure.ResourceManager.Workloads.Models.SapHealthState? Health { get { throw null; } }
+        public string Hostname { get { throw null; } }
+        public string IPAddress { get { throw null; } }
+        public long? Port { get { throw null; } }
+    }
+    public partial class ErrorDefinition
```
Comment (fengzhou-msft) [inline sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs:697]: Remove `Definition` suffix and probably add a prefix like `Workloads`?

File: sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs [hunk 3]
```diff
@@ -0,0 +1,1742 @@
...
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public override bool Equals(object obj) { throw null; }
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public override int GetHashCode() { throw null; }
+        public static bool operator ==(Azure.ResourceManager.Workloads.Models.HAEnabled left, Azure.ResourceManager.Workloads.Models.HAEnabled right) { throw null; }
+        public static implicit operator Azure.ResourceManager.Workloads.Models.HAEnabled (string value) { throw null; }
+        public static bool operator !=(Azure.ResourceManager.Workloads.Models.HAEnabled left, Azure.ResourceManager.Workloads.Models.HAEnabled right) { throw null; }
+        public override string ToString() { throw null; }
+    }
+    public partial class HanaDbProviderInstanceProperties : Azure.ResourceManager.Workloads.Models.ProviderSpecificProperties
```
Comment (fengzhou-msft) [inline sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs:775]: HanaDBProviderInstanceProperties

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/HighAvailabilitySoftwareConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            {
+                throw new ArgumentNullException(nameof(fencingClientPassword));
+            }
+
+            FencingClientId = fencingClientId;
+            FencingClientPassword = fencingClientPassword;
+        }
+
+        /// <summary> The fencing client id. </summary>
+        public string FencingClientId { get; set; }
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/HighAvailabilitySoftwareConfiguration.cs:35]: A `Guid`?

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/RedisCacheFamily.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            _value = value ?? throw new ArgumentNullException(nameof(value));
+        }
+
+        private const string CValue = "C";
+        private const string PValue = "P";
+
+        /// <summary> C. </summary>
+        public static RedisCacheFamily C { get; } = new RedisCacheFamily(CValue);
+        /// <summary> P. </summary>
+        public static RedisCacheFamily P { get; } = new RedisCacheFamily(PValue);
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/RedisCacheFamily.cs:31]: It's weird to use single letter, can we change it?

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SapNetWeaverProviderInstanceProperties.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        /// <summary> Gets or sets the list of HostFile Entries. </summary>
+        public IList<string> SapHostFileEntries { get; }
+        /// <summary> Gets or sets the SAP user name. </summary>
+        public string SapUsername { get; set; }
+        /// <summary> Sets the SAP password. </summary>
+        public string SapPassword { get; set; }
+        /// <summary> Gets or sets the key vault URI to secret with the SAP password. </summary>
+        public Uri SapPasswordUri { get; set; }
+        /// <summary> Gets or sets the SAP Client ID. </summary>
+        public string SapClientId { get; set; }
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SapNetWeaverProviderInstanceProperties.cs:66]: Should be `Guid`?

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SapProductType.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            _value = value ?? throw new ArgumentNullException(nameof(value));
+        }
+
+        private const string ECCValue = "ECC";
+        private const string S4HanaValue = "S4HANA";
+        private const string OtherValue = "Other";
+
+        /// <summary> ECC. </summary>
+        public static SapProductType ECC { get; } = new SapProductType(ECCValue);
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SapProductType.cs:30]: `Ecc`?

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SkuDefinition.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        /// <summary> This field is required to be implemented by the Resource Provider if the service has more than one tier, but is not required on a PUT. </summary>
+        public string Tier { get; }
+        /// <summary> The SKU size. When the name field is the combination of tier and some other value, this would be the standalone code. </summary>
+        public string Size { get; }
+        /// <summary> If the service has different generations of hardware, for the same SKU, then that can be captured here. </summary>
+        public string Family { get; }
+        /// <summary> If the service has different kinds of hardware, for the same SKU, then that can be captured here. </summary>
+        public string Kind { get; }
+        /// <summary> List of locations where this SKU is available. </summary>
+        public IReadOnlyList<string> Locations { get; }
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SkuDefinition.cs:77]: Should the type be `AzureLocation`?

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/WordpressVersions.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+        /// <summary> 5.4.3. </summary>
+        public static WordpressVersions Five43 { get; } = new WordpressVersions(Five43Value);
+        /// <summary> 5.4.2. </summary>
+        public static WordpressVersions Five42 { get; } = new WordpressVersions(Five42Value);
+        /// <summary> 5.4.1. </summary>
+        public static WordpressVersions Five41 { get; } = new WordpressVersions(Five41Value);
+        /// <summary> 5.4. </summary>
+        public static WordpressVersions Five4 { get; } = new WordpressVersions(Five4Value);
+        /// <summary> Determines if two <see cref="WordpressVersions"/> values are the same. </summary>
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/WordpressVersions.cs:38]: These properties name looks weird, I would suggest using `Version5_4_0`, `Version5_4_1`, `Version5_4_2` and `Version5_4_3`, you can take a look at [this change](https://github.com/azure/azure-sdk-for-net/blob/main/sdk/cdn/Azure.ResourceManager.Cdn/src/Customization/Models/SslProtocol.cs#L15-L25) to see how to use customization to implement this.

File: sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs (hunk 1 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs:515]: Our currently rule is `Vm`

File: sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs [hunk 10]
```diff
@@ -0,0 +1,1742 @@
...
+    public partial class DatabaseVmDetails
+    {
+        internal DatabaseVmDetails() { }
+        public Azure.ResourceManager.Workloads.Models.SapVirtualInstanceStatus? Status { get { throw null; } }
+        public Azure.Core.ResourceIdentifier VirtualMachineId { get { throw null; } }
+    }
+    public partial class Db2ProviderInstanceProperties : Azure.ResourceManager.Workloads.Models.ProviderSpecificProperties
+    {
+        public Db2ProviderInstanceProperties() { }
+        public string DbName { get { throw null; } set { } }
```
[the remark this answers is not in the document]
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs:577]: fixed

File: sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs (hunk 2 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs:697]: Rename to `SapVirtualInstanceErrorDetail`

File: sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs (hunk 3 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/api/Azure.ResourceManager.Workloads.netstandard2.0.cs:775]: fixed

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/DatabaseProfile.cs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        /// <summary> Database server name. </summary>
+        public string ServerName { get; set; }
+        /// <summary> Database version. </summary>
+        public string Version { get; set; }
+        /// <summary> The name of the server SKU, e.g. Standard_D32s_v4. </summary>
+        public string Sku { get; set; }
+        /// <summary> Tier of the server SKU. </summary>
+        public DatabaseTier Tier { get; set; }
+        /// <summary> Whether to enable HA for the server. </summary>
+        public HAEnabled? HaEnabled { get; set; }
```
[the remark this answers is not in the document]
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/DatabaseProfile.cs:73]: fixed

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/EnqueueReplicationServerType.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            _value = value ?? throw new ArgumentNullException(nameof(value));
+        }
+
+        private const string EnqueueReplicator1Value = "EnqueueReplicator1";
+        private const string EnqueueReplicator2Value = "EnqueueReplicator2";
+
+        /// <summary> EnqueueReplicator1. </summary>
+        public static EnqueueReplicationServerType EnqueueReplicator1 { get; } = new EnqueueReplicationServerType(EnqueueReplicator1Value);
+        /// <summary> EnqueueReplicator2. </summary>
```
[the remark this answers is not in the document]
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/EnqueueReplicationServerType.cs:30]: will check with service team

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/HighAvailabilitySoftwareConfiguration.cs (hunk 4 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/HighAvailabilitySoftwareConfiguration.cs:35]: not `guid`

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SapNetWeaverProviderInstanceProperties.cs (hunk 6 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SapNetWeaverProviderInstanceProperties.cs:66]: not `guid`

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SapProductType.cs (hunk 7 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SapProductType.cs:30]: fixed

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/RedisCacheFamily.cs (hunk 5 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/RedisCacheFamily.cs:31]: It's Redis concept, I don't think we need to rename it

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SkuDefinition.cs (hunk 8 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/SkuDefinition.cs:77]: fixed

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/OperationStatusResult.cs [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+using System;
+using System.Collections.Generic;
+using Azure;
+using Azure.Core;
+
+namespace Azure.ResourceManager.Workloads.Models
+{
+    /// <summary> The current status of an async operation. </summary>
+    public partial class OperationStatusResult
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/OperationStatusResult.cs:16]: Can we rename it to `WorkloadsOperationStatusResult`? I saw it was uesd in different `delete`, `start` or `stop` method, maybe add a RP prefix is better.

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/OSImageSku.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+            _value = value ?? throw new ArgumentNullException(nameof(value));
+        }
+
+        private const string Eighteen04LTSValue = "18.04-LTS";
+        private const string Sixteen04LTSValue = "16.04-LTS";
+
+        /// <summary> 18.04-LTS. </summary>
+        public static OSImageSku Eighteen04LTS { get; } = new OSImageSku(Eighteen04LTSValue);
+        /// <summary> 16.04-LTS. </summary>
+        public static OSImageSku Sixteen04LTS { get; } = new OSImageSku(Sixteen04LTSValue);
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/OSImageSku.cs:31]: These names are not good, I would suggest:
```suggestion
        public static OSImageSku Ubuntu18_04Lts { get; } = new OSImageSku(Eighteen04LTSValue);
        /// <summary> 16.04-LTS. </summary>
        public static OSImageSku Ubuntu16_04Lts { get; } = new OSImageSku(Sixteen04LTSValue);
```

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/PhpVersion.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        private const string Seven2Value = "7.2";
+        private const string Seven3Value = "7.3";
+        private const string Seven4Value = "7.4";
+
+        /// <summary> 7.2. </summary>
+        public static PhpVersion Seven2 { get; } = new PhpVersion(Seven2Value);
+        /// <summary> 7.3. </summary>
+        public static PhpVersion Seven3 { get; } = new PhpVersion(Seven3Value);
+        /// <summary> 7.4. </summary>
+        public static PhpVersion Seven4 { get; } = new PhpVersion(Seven4Value);
```
Comment (Yao725) [inline sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/PhpVersion.cs:34]: Same suggestion as above.
```suggestion
        public static PhpVersion Php7_2 { get; } = new PhpVersion(Seven2Value);
        /// <summary> 7.3. </summary>
        public static PhpVersion Php7_3 { get; } = new PhpVersion(Seven3Value);
        /// <summary> 7.4. </summary>
        public static PhpVersion Php7_4 { get; } = new PhpVersion(Seven4Value);
```

**Review by Yao725 — APPROVED** [review]
LGTM, left three comments for improvement.

File: sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/PhpVersion.cs (hunk 15 above)
Comment (ArthurMa1978) [reply sdk/workloads/Azure.ResourceManager.Workloads/src/Generated/Models/PhpVersion.cs]: customized to `V7.2`, `V7.3`, `V7.4`

**Pull request merged**